plugin/trino-kafka-event-listener/src/main/java/io/trino/plugin/eventlistener/kafka/KafkaEventPublisher.java

```diff
@@ -56,21 +75,81 @@ public KafkaEventPublisher(KafkaEventListenerConfig config, KafkaProducerFactory
-        LOG.info("Creating Kafka publisher (SSL=%s) for topics: %s/%s with excluded fields: %s and kafka config overrides: %s",
-                producerFactory instanceof SSLKafkaProducerFactory, createdTopic, completedTopic, config.getExcludedFields(), configOverrides);
-        kafkaProducer = producerFactory.producer(configOverrides);
-        checkConnectivityToBrokers(config.getPublishCreatedEvent() ? createdTopic : completedTopic, config.getRequestTimeout().toMillis());
+        kafkaProducerSupplier = () -> {
+            Map<String, String> configOverrides = config.getKafkaClientOverrides();
+            LOG.info("Creating Kafka publisher (SSL=%s) for topics: %s/%s with excluded fields: %s and kafka config overrides: %s",
+                    producerFactory instanceof SSLKafkaProducerFactory, createdTopic, completedTopic, config.getExcludedFields(), configOverrides);
+            KafkaProducer<String, String> kafkaProducer = producerFactory.producer(configOverrides);
+            try {
+                checkConnectivityToBrokers(config.getPublishCreatedEvent() ? createdTopic : completedTopic, config.getRequestTimeout().toMillis());
+            }
+            catch (Throwable e) {
+                try {
+                    kafkaProducer.close();
+                }
+                catch (Throwable closeException) {
+                    e.addSuppressed(closeException);
+                }
+                throw new RuntimeException("Cannot connect to kafka broker", e);
+            }
+            return kafkaProducer;
+        };
         kafkaRecordBuilder = new KafkaRecordBuilder(createdTopic, completedTopic, splitCompletedTopic, config.getExcludedFields(), metadataProvider(config));
         LOG.info("Successfully created Kafka publisher.");
     }
 
+    private static final java.time.Duration KAFKA_CONNECT_INTERVAL = Duration.of(10, SECONDS);
+
+    @PostConstruct
+    public void createKafkaProducer()
+    {
+        // try to connect to Kafka eagerly every now and then
+        connectExecutor.execute(() -> {
+            while (true) {
+                try {
+                    getProducer();
+                    // success
+                    return;
+                }
+                catch (Exception e) {
+                    LOG.error("Could not create Kafka producer", e);
+                    try {
+                        sleep(KAFKA_CONNECT_INTERVAL.toMillis());
+                    }
+                    catch (InterruptedException ex) {
+                        Thread.currentThread().interrupt();
+                        // finishing work
+                        return;
+                    }
+                }
+            }
+        });
+    }
+
+    private synchronized Producer<String, String> getProducer()
+    {
+        if (kafkaProducer != null) {
+            return kafkaProducer;
+        }
+        if (lastKafkaProducerTime.plus(KAFKA_CONNECT_INTERVAL).isAfter(now())) {
```

Comment:
What does it solve really? I think Kafka listener either should be there or not. There is `kafka-event-listener.terminate-on-initialization-failure` if we don't want Trino startup to fail in case listener creation fails

Reply:
Oh - I did not know that there is this mechanism. It solves problem of not terminating trino if initialization fails, yeah. But do you want it to not try to connect again when intialization failed? Seems like it should try until it establishes connection. Similar like it would if you loose connection to Kafka while Trino is running.